src/Crypto/Macaroon/Verifier/Internal.hs

```diff
@@ -14,54 +15,39 @@ Portability : portable
 -}
 module Crypto.Macaroon.Verifier.Internal where
 
-import           Control.Applicative
-import           Control.Monad
-import           Control.Monad.IO.Class
+import           Control.Arrow            ((&&&))
 import           Crypto.Hash
 import           Data.Bool
 import           Data.Byteable
 import qualified Data.ByteString          as BS
-import           Data.Either
-import           Data.Either.Validation
 import           Data.Foldable
+import           Data.List.NonEmpty       (NonEmpty, nonEmpty)
 import           Data.Maybe
-import           Data.Monoid hiding ((<>))
-import           Data.Semigroup
+import           Data.Text                (Text)
 
 import           Crypto.Macaroon.Internal
 
 
 -- | Type representing the result of a validator
 data VerifierResult = Verified -- ^ The caveat is correctly parsed and verified
-                    | Refused ValidationError -- ^ The caveat is refused (Either a parse error or a validation error
+                    | Refused VerifierError -- ^ The caveat is refused
                     | Unrelated -- ^ The given verifier does not verify the caveat
                     deriving (Show, Eq)
 
--- | Type representing different validation errors.
--- Only 'ParseError' and 'ValidatorError' are exported, @SigMismatch@ and
--- @NoVerifier@ are used internally and should not be used by the user
-data ValidationError = SigMismatch -- ^ Signatures do not match
-                     | NoVerifier -- ^ No verifier can handle a given caveat
-                     | ParseError String -- ^ A verifier had a parse error
-                     | ValidatorError String -- ^ A verifier failed
-                     deriving (Show,Eq)
-
+-- | Type representing an error returned by a verifier (on a caveat it's supposed to handle)
+data VerifierError = ParseError Text -- ^ The caveat couldn't be parsed
+                   | VerifierError Text -- ^ The verifier understood the caveat but couldn't satisfy it.
+                   deriving (Show, Eq)
 
--- | The 'Semigroup' instance is written so @SigMismatch@ is an annihilator
-instance Semigroup ValidationError where
-    NoVerifier <> e = e
-    e <> NoVerifier = e
-    SigMismatch <> _ = SigMismatch
-    _ <> SigMismatch = SigMismatch
-    (ValidatorError e) <> (ParseError _) = ValidatorError e
-    (ParseError _) <> (ValidatorError e) = ValidatorError e
-    (ValidatorError e1) <> (ValidatorError e2) = ValidatorError $ e1 <> " " <> e2
-    (ParseError e1) <> (ParseError e2) = ParseError $ e1 <> " " <> e2
+-- | Type alias for a caveat that was not discharged
+--   an empty list means that no verifiers were related to the caveat
+type OutstandingCaveat = (Caveat, [VerifierError])
```

Comment:
I'm not sure about the name. Maybe something closer to the comment, like _NonDischargedCaveat_ ?
```suggestion
type NonDischargedCaveat = (Caveat, [VerifierError])
```

Reply:
i can make that change if you want

Reply:
I don't know. I can understand why you chose that name, but at the same time, as a non native english speaker, I didn't fully grasped the meaning just by looking at the name.

Reply:
let's make it more obvisous then.
Either `NonDischargedCaveat` or maybe `RemainingCaveat` ?